hera_cal/vis_clean.py

```diff
@@ -0,0 +1,471 @@
+                gain_convention = 'divide'
+            elif gain_convention == 'divide':
+                gain_convention = 'multiply'
+        apply_cal.calibrate_in_place(self.data, cal_gains, self.flags, cal_flags, gain_convention=gain_convention)
+
+    def read(self, **read_kwargs):
+        """
+        Read visibility data from self.hd and attach data and/or metadata to self.
+
+        Args:
+            read_kwargs : dictionary
+                Keyword arguments to pass to HERAData.read().
+        """
+        # read data
+        self.hd.read(return_data=False, **read_kwargs)
+
+        # attach data
+        self.attach_data()
+
+    def write_data(self, data, filename, overwrite=False, flags=None, nsamples=None, filetype='uvh5',
+                   partial_write=False, add_to_history=None, verbose=True, **kwargs):
+        """
+        Write data attached to object to file. If data or flags are fed as DataContainers,
+        create a new HERAData with those data and write that to file. Can only write
+        data that has associated metadata in the self.hd HERAData object.
+
+        Args:
+            data : DataContainer, holding complex visibility data to write to disk.
+            filename : string, output filepath
+            overwrite : bool, if True, overwrite output file if it exists
+            flags : DataContainer, boolean flag arrays to write to disk with data.
+            nsamples : DataContainer, float nsample arrays to write to disk with data.
+            filetype : string, output filetype. ['miriad', 'uvh5', 'uvfits'] supported.
+            partial_write : bool, if True, begin (or continue) a partial write to
+            the output filename and store file descriptor in self.hd._writers.
+            add_to_history : string, string to append to hd history.
+            kwargs : additional attributes to update before write to disk.
+        """
+        # get common keys
+        keys = [k for k in self.hd.get_antpairpols() if data.has_key(k)]
+        if flags is not None:
+            keys = [k for k in keys if flags.has_key(k)]
+        if nsamples is not None:
+            keys = [k for k in keys if nsamples.has_key(k)]
+
+        # select out a copy of hd
+        hd = self.hd.select(bls=keys, inplace=False)
+
+        # update HERAData
+        hd.update(data=data, flags=flags, nsamples=nsamples)
+ 
+        # add history
+        if add_to_history is not None:
+            hd.history = "{}\n{}".format(hd.history, add_to_history)
+
+        # update other kwargs
+        for attribute, value in kwargs.items():
+            hd.__setattr__(attribute, value)
+
+        # write to disk
+        if filetype == 'miriad':
+            hd.write_miriad(filename, clobber=overwrite)
+        elif filetype == 'uvh5':
+            if partial_write:
+                hd.partial_write(filename, clobber=overwrite, inplace=True)
+                self.hd._writers.update(hd._writers)
```

Comment:
Does this touch the `self.hd` object, or add to the amount of data in memory?

Reply:
Was able to confirm that this does not touch the `self.hd` object, and that what is stored in `self.hd._writers` is not a reference to `hd`, therefore `hd` should be destroyed upon function exit...